airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -117,70 +229,120 @@ def _evaluate_trigger_rule(
         :param dep_context: The current dependency context.
         :param session: Database session.
         """
-        from airflow.models.abstractoperator import NotMapped
-        from airflow.models.expandinput import NotFullyPopulated
         from airflow.models.operator import needs_expansion
         from airflow.models.taskinstance import TaskInstance
 
         task = ti.task
-        upstream_tasks = {t.task_id: t for t in task.upstream_list}
-        trigger_rule = task.trigger_rule
 
-        @functools.lru_cache
-        def _get_expanded_ti_count() -> int:
-            """Get how many tis the current task is supposed to be expanded into.
+        indirect_setups = {k: v for k, v in relevant_setups.items() if k not in task.upstream_task_ids}
+        finished_upstream_tis = (
+            x
+            for x in dep_context.ensure_finished_tis(ti.get_dagrun(session), session)
+            if self._is_relevant_upstream(
+                ti=ti,
+                upstream=x,
+                relevant_ids=indirect_setups.keys(),
+                session=session,
+            )
+        )
+        upstream_states = _UpstreamTIStates.calculate(finished_upstream_tis)
 
-            This extra closure allows us to query the database only when needed,
-            and at most once.
-            """
-            return task.get_mapped_ti_count(ti.run_id, session=session)
+        # all of these counts reflect indirect setups which are relevant for this ti
+        success = upstream_states.success
+        skipped = upstream_states.skipped
+        failed = upstream_states.failed
+        upstream_failed = upstream_states.upstream_failed
+        removed = upstream_states.removed
 
-        @functools.lru_cache
-        def _get_relevant_upstream_map_indexes(upstream_id: str) -> int | range | None:
-            """Get the given task's map indexes relevant to the current ti.
+        # Optimization: Don't need to hit the database if all upstreams are
+        # "simple" tasks (no task or task group mapping involved).
+        if not any(needs_expansion(t) for t in indirect_setups.values()):
+            upstream = len(indirect_setups)
+        else:
+            task_id_counts = session.execute(
+                select(TaskInstance.task_id, func.count(TaskInstance.task_id))
+                .where(TaskInstance.dag_id == ti.dag_id, TaskInstance.run_id == ti.run_id)
+                .where(
+                    or_(
+                        *self._iter_upstream_conditions(
+                            ti=ti, relevant_tasks=indirect_setups, session=session
+                        )
+                    )
+                )
+                .group_by(TaskInstance.task_id)
+            ).all()
+            upstream = sum(count for _, count in task_id_counts)
```

Comment:
Do we really need an expensive group by here?
```suggestion
            upstream = session.execute(
                select(func.count("*"))
                .where(TaskInstance.dag_id == ti.dag_id, TaskInstance.run_id == ti.run_id)
                .where(
                    or_(
                        *self._iter_upstream_conditions(
                            ti=ti, relevant_tasks=indirect_setups, session=session
                        )
                    )
                )
            ).scalar_one()
```

Reply:
i would rather leave this to future.